### ISSUE TITLE
Figure out nice semantics for attaching handlers per row

### ISSUE THREAD

**Issue by bebraw** [bebraw/reactabular#56]
To quote @utdemir at https://github.com/bebraw/reactabular/commit/2837cc9c6de070fb3f19831a81ab092dc6f6ae1a#commitcomment-11556460 .

> Well, reactabular looks pretty configurable, thanks for building it like that :). My only small nitpick was when I tried to put an onclick() action to a row, instead of a column; I couldn't figure out a way, and had to put the same onClick actions to every column and dispatched via rowIndex. I'd prefer to put an onclick action to every data, or every row etc.

```javascript
  [{
        property: 'displayName',
        header: 'Name',
        cell: (value, data, rowIndex) =>
          <span
            onClick={() => this.navigate(data[rowIndex].id)}
            className={this.state.threatId == data[rowIndex].id ? "active" : "" }
            >
            {value}
          </span>
      }, {
        property: 'category',
        header: 'Category',
        cell: (value, data, rowIndex) =>
          <span
            onClick={() => this.navigate(data[rowIndex].id)}
            className={this.state.id == data[rowIndex].id ? "active" : "" }
            >
            {data[rowIndex].category[0]}
          </span>
      }];
```

We probably want to expose `tr`'s somehow to make this happen. Need to figure out where/how to do that most naturally.

**Comment by utdemir** [comment]
Also, the same applies to `className={this.state.threatId == data[rowIndex].id ? "active" : "" }` line, I want to make the whole row `active`, exposing tr's would solve that too.

Thanks again.

**Comment by utdemir** [comment]
Hey, if you're able to figure out a nice API, I'm willing to implement this feature.

**Comment by bebraw** [comment]
@utdemir Cool. It doesn't seem like that big of a change to make. Here's a rough spec:

```javascript
<Table row={{onClick: () => ..., className: 'row', ...}} ... />
```

On implementation level you would apply that on `tr`s (both header and content) like `<tr {...this.props.row}>`.

I have a feeling that could be a starting point. Please tweak to fit your purposes. Also work against `dev` branch if possible. That will make it easier to deal with the merge.

**Comment by eviltoylet** [comment]
The introduction of this knob introduces yet another set of complexities that need to be addressed. I'd prefer to avoid it.

As Utku demonstrated above, the library already has the power to express row handlers, albeit a bit cumbersome. What if we were to provide some helpers?
- apply the same cell formatter to all columns
- have a decorator which will apply a formatter to a given column/row

**Comment by bebraw** [comment]
> The introduction of this knob introduces yet another set of complexities that need to be addressed. I'd prefer to avoid it.

That's definitely true!

As you suggested some sort of helpers could work. I started thinking it from another direction. Currently we have nice semantics for dealing with columns. Data is per row. And that's where the weakness is. What if we agreed on some convention that allowed you to put row `props` within the data? I mean something along this:

```javascript

var data = this.state.data;

// relying on convention here. we'll hook into _getProps
// about a line of code at library side
// something along {...row._getProps(i)} for tr
data = data.map((d) => {
  d._getProps = (rowIndex) => {
    onClick: () => {console.log('row clicked');},
    className: rowIndex === this.state.id && 'selected',
  };

  return d;
});

return <Table data={data} ... />
```

This approach would avoid that extra prop while allowing you to modify props of `tbody` `tr`'s.

**Comment by utdemir** [comment]
@eviltoylet 

> As Utku demonstrated above, the library already has the power to express row handlers, albeit a bit cumbersome.

I have to disagree with that. For example; just this afternoon I wanted to highlight to a row. At first, I put a background color to every cell on that row:

* The paddings between two cells is empty and need many CSS tricks concatenate them. 
* Even if the the cells look together and have the same height, the contents had different heights. I had to play with CSS a little more to give the same height to every cell in a row.

When I could have just put this:

    tr .active {
      background-color: #deadbeef 
    }

It's not a little bit cumbersome, it's my only headache with reactabular :).

@bebraw Your solution also works for me nicely, but I'd prefer to keep my data closer to original and give rendering logic to reactabular via somewhere else(like your initial suggestion).

Example:
```
  <Table data={} columns={} rowProps={(data, rowIndex) => { className: rowIndex % ? "a" : "b", onClick: => console.log("clicked: " + data) }
```

**Comment by bebraw** [comment]
I have a proof of concept at `dev` branch. I went with the `row` prop as it seemed like the easiest way to get this resolved. It's true something could have been built on top of the current cell API but this was the easiest way to get forward. We can revisit this later if it starts to feel like a kludge or causes some complications elsewhere.

@utdemir Can you have a look? If this is fine, I'll push an intermediate release with the feature in it.

**Comment by eviltoylet** [comment]
@utdemir agreed. Thanks for presenting like that :)
@bebraw great work as always!

**Comment by utdemir** [comment]
@bebraw Sorry for not answering, I didn't had time to look into it.

Thanks for this great library!

**Comment by bebraw** [comment]
@utdemir No probs! I just published `0.6.4` that contains the feature. :+1:
